File: src/predictfun/coordinator.py
from __future__ import annotations

from dataclasses import dataclass, field
from decimal import Decimal

from .accounts import GlobalTradingPaused, PolymarketAccountState, PredictAccountRotator, PredictAccountState
from .adapters import PolymarketAdapter, PredictAdapter
from .alerts import Alert, AlertSink
from .audit import AuditEvent, AuditSink
from .compliance import ComplianceResult
from .discovery import MarketDiscovery
from .engine import ArbEngine, BookBundle, FeeRates
from .execution import ExecutionPlan, ExecutionReport, TwoLegExecutor
from .matching import MarketMatch
from .models import OrderRequest, OrderSide, OrderType, Outcome, Venue
from .risk_manager import RiskDecision, RiskManager
from .scanner import BtcMarketScanner, ScanResult


@dataclass(frozen=True)
class CoordinatorResult:
    scan: ScanResult
    decisions: tuple[RiskDecision, ...]
    reports: tuple[ExecutionReport, ...]
    paused: bool = False
    pause_reason: str | None = None


class ExecutionCoordinator:
    def __init__(
        self,
        *,
        predict_adapter: PredictAdapter,
        polymarket_adapter: PolymarketAdapter,
        predict_rotator: PredictAccountRotator,
        polymarket_account: PolymarketAccountState,
        discovery: MarketDiscovery | None = None,
        scanner: BtcMarketScanner | None = None,
        engine: ArbEngine | None = None,
        risk_manager: RiskManager | None = None,
        executor: TwoLegExecutor | None = None,
        audit: AuditSink | None = None,
        alerts: AlertSink | None = None,
        compliance: ComplianceResult | None = None,
        dry_run: bool = True,
        live_trading_enabled: bool = False,
        fee_rates: FeeRates | None = None,
    ) -> None:
        self.predict_adapter = predict_adapter
        self.polymarket_adapter = polymarket_adapter
        self.predict_rotator = predict_rotator
        self.polymarket_account = polymarket_account
        self.discovery = discovery or MarketDiscovery(predict=predict_adapter, polymarket=polymarket_adapter)
        self.scanner = scanner or BtcMarketScanner()
        self.engine = engine or ArbEngine()
        self.risk_manager = risk_manager or RiskManager(engine=self.engine)
        self.executor = executor or TwoLegExecutor(
            predict_client=predict_adapter,
            polymarket_client=polymarket_adapter,
        )
        self.audit = audit
        self.alerts = alerts
        self.compliance = compliance or ComplianceResult(ok=True, reasons=())
        self.dry_run = dry_run
        self.live_trading_enabled = live_trading_enabled
        self.fee_rates = fee_rates or FeeRates(predict_fee_rate_bps=0, polymarket_fee_rate_bps=0)

    def run_once(self) -> CoordinatorResult:
        discovered = self.discovery.discover()
        self._audit("market_discovery", "discovered BTC markets", {
            "predict": len(discovered.predict_markets),
            "polymarket": len(discovered.polymarket_markets),
        })

        scan = self.scanner.scan(discovered.predict_markets, discovered.polymarket_markets)
        self._audit("market_matcher", "strict matcher completed", {
            "accepted": len(scan.accepted),
            "rejected": len(scan.rejected),
        })

        decisions: list[RiskDecision] = []
        reports: list[ExecutionReport] = []
        for match in scan.accepted:
            books = self._books_for(match)
            try:
                decision = self.risk_manager.choose_trade(
                    books=books,
                    fee_rates=self.fee_rates,
                    predict_rotator=self.predict_rotator,
                    polymarket_account=self.polymarket_account,
                )
            except GlobalTradingPaused as exc:
                reason = str(exc)
                self._audit("risk_manager", "global pause", {"reason": reason})
                self._alert("warning", reason)
                return CoordinatorResult(scan=scan, decisions=tuple(decisions), reports=tuple(reports), paused=True, pause_reason=reason)

            decisions.append(decision)
            if not decision.accepted or decision.sizing is None or decision.sizing.quote is None or decision.predict_account_id is None:
                self._audit("risk_manager", "no executable hedge", {"reasons": decision.reasons})
                continue

            plan = self._plan_for(match, decision)
            report = self.executor.execute(plan)
            reports.append(report)
            self._audit("execution", "execution report", {
                "status": report.status,
                "pause_opening": report.pause_opening,
                "pause_reason": report.pause_reason,
            })
            self._settle_predict_account(decision.predict_account_id, report)

            if report.pause_opening:
                reason = report.pause_reason or "execution paused new openings"
                self._alert("warning", reason)
                return CoordinatorResult(scan=scan, decisions=tuple(decisions), reports=tuple(reports), paused=True, pause_reason=reason)

        return CoordinatorResult(scan=scan, decisions=tuple(decisions), reports=tuple(reports))

    def _books_for(self, match: MarketMatch) -> BookBundle:
        return BookBundle(
            predict_yes=self.predict_adapter.get_orderbook(match.predict, Outcome.YES),
            predict_no=self.predict_adapter.get_orderbook(match.predict, Outcome.NO),
            polymarket_yes=self.polymarket_adapter.get_orderbook(match.polymarket, Outcome.YES),
            polymarket_no=self.polymarket_adapter.get_orderbook(match.polymarket, Outcome.NO),
        )

    def _plan_for(self, match: MarketMatch, decision: RiskDecision) -> ExecutionPlan:
        assert decision.sizing is not None
        assert decision.sizing.quote is not None
        assert decision.predict_account_id is not None
        combo = decision.sizing.combo
        predict_outcome = self.risk_manager.predict_outcome_for(combo)
        polymarket_outcome = self.risk_manager.polymarket_outcome_for(combo)
        quote = decision.sizing.quote
        predict_order = OrderRequest(
            venue=Venue.PREDICT,
            market_id=match.predict.venue_market_id,
            outcome=predict_outcome,
            side=OrderSide.BUY,
            order_type=OrderType.FOK,
            shares=quote.shares,
            limit_price=quote.predict_leg.fill.worst_price or quote.predict_leg.fill.average_price,
            account_id=decision.predict_account_id,
            client_order_id=f"predict-{combo.value}-{match.predict.venue_market_id}",
        )
        polymarket_order = OrderRequest(
            venue=Venue.POLYMARKET,
            market_id=match.polymarket.venue_market_id,
            outcome=polymarket_outcome,
            side=OrderSide.BUY,
            order_type=OrderType.FOK,
            shares=quote.shares,
            limit_price=quote.polymarket_leg.fill.worst_price or quote.polymarket_leg.fill.average_price,
            account_id=self.polymarket_account.account_id,
            client_order_id=f"poly-{combo.value}-{match.polymarket.venue_market_id}",
        )
        return ExecutionPlan(
            market_match=match,
            predict_order=predict_order,
            polymarket_order=polymarket_order,
            quote=quote,
            dry_run=self.dry_run,
            live_trading_enabled=self.live_trading_enabled,
            compliance=self.compliance,
            predict_account=self._account(decision.predict_account_id),
            polymarket_account=self.polymarket_account,
        )

    def _settle_predict_account(self, account_id: str, report: ExecutionReport) -> None:
        if report.predict_held_position is not None:
            self.predict_rotator.mark_held(account_id, report.predict_held_position)
        else:
            self.predict_rotator.release(account_id)

    def _account(self, account_id: str) -> PredictAccountState | None:
        for account in self.predict_rotator.accounts:
            if account.account_id == account_id:
                return account
        return None

    def _audit(self, event_type: str, message: str, data: dict[str, object] | None = None) -> None:
        if self.audit is not None:
            self.audit.record(AuditEvent(event_type=event_type, message=message, data=dict(data or {})))

    def _alert(self, level: str, message: str) -> None:
        if self.alerts is not None:
            self.alerts.send(Alert(level=level, message=message))

File: src/execution/coordinator.ts
import {
  PredictAccountRotator,
  ensurePolymarketCanOpen,
  isReady,
  type HeldPosition,
  type PolymarketAccountState
} from "../accounts/rotator.js";
import { type PolymarketAdapter, type PredictAdapter } from "../adapters/contracts.js";
import { alert, type AlertSink } from "../alerts/alertSink.js";
import { auditEvent, type AuditSink } from "../audit/auditSink.js";
import { ArbEngine, type FeeRates } from "../arb/engine.js";
import { MarketDiscovery } from "../discovery/marketDiscovery.js";
import { type BinaryMarketSpec, type OrderRequest, type OrderResult } from "../domain/models.js";
import { d, ZERO, type D } from "../domain/money.js";
import { isEligibleShortWindowBtcMarket, type ShortWindowFilterConfig } from "../core/short-window-market-filter.js";
import { type NormalizedMarket } from "../core/types.js";
import { BtcMarketScanner } from "../markets/scanner.js";
import { emitMonitorEvent, orderSubmitted, tradeRejected, botPaused } from "../monitoring/observability.js";
import { type MetricsSink } from "../monitoring/metrics.js";
import { StrategyEngine } from "../strategy/strategy-engine.js";
import { type StrategyConfig } from "../strategy/types.js";
import {
  buildResidualRescueOrder,
  confirmOrderFill,
  reconcileExecutionSafety,
  validateConservativeOrderPlan,
  type ExecutionSafetyOutcome,
  type ExecutionSafetyPolicy
} from "./safety.js";

export interface CoordinatorResult {
  paused: boolean;
  pauseReason?: string;
  executed: number;
}

const DEFAULT_STRATEGY_CONFIG: StrategyConfig = {
  strategyMode: "pure_arbitrage",
  hedgeEnabled: false,
  maxNetExposureUsd: ZERO,
  maxPredictUsagePct: d("0.30"),
  minProfitAfterHedgeFee: ZERO
};

export interface ShortWindowExecutionGuardInput {
  market: NormalizedMarket;
  nowMs: number;
  cfg: ShortWindowFilterConfig;
  staleBookMs: number;
  predictBookTs: number;
  polymarketBookTs: number;
  expectedProfitUsd: D;
  expectedProfitPerShare: D;
  orderType: OrderRequest["orderType"];
}

export interface ShortWindowExecutionGuardResult {
  ok: boolean;
  reasons: readonly string[];
  secondsToClose?: number;
}

export function validateShortWindowExecutionGuard(input: ShortWindowExecutionGuardInput): ShortWindowExecutionGuardResult {
  const reasons: string[] = [];
  const eligibility = isEligibleShortWindowBtcMarket(input.market, input.nowMs, input.cfg);
  if (!eligibility.approved) reasons.push(`REJECT_${eligibility.reason}`);
  if (input.nowMs - input.predictBookTs > input.staleBookMs || input.nowMs - input.polymarketBookTs > input.staleBookMs) {
    reasons.push("REJECT_STALE_BOOK");
  }
  if (input.expectedProfitUsd.lte(0) || input.expectedProfitPerShare.lte(0)) {
    reasons.push("REJECT_NO_PROFIT_AFTER_RECHECK");
  }
  if (input.orderType === "MARKET") {
    reasons.push("REJECT_NAKED_MARKET_ORDER");
  }

  return {
    ok: reasons.length === 0,
    reasons: [...new Set(reasons)],
    secondsToClose: eligibility.approved ? eligibility.secondsToClose : undefined
  };
}

export class ExecutionCoordinator {
  constructor(
    private readonly deps: {
      predict: PredictAdapter;
      polymarket: PolymarketAdapter;
      predictRotator: PredictAccountRotator;
      polymarketAccount: PolymarketAccountState;
      engine: ArbEngine;
      feeRates: FeeRates;
      audit?: AuditSink;
      alerts?: AlertSink;
      metrics?: MetricsSink;
      dryRun: boolean;
      liveTradingEnabled: boolean;
      safetyPolicy?: ExecutionSafetyPolicy;
      strategyConfig?: StrategyConfig;
      strategyEngine?: StrategyEngine;
    }
  ) {}

  async runOnce(): Promise<CoordinatorResult> {
    const discovery = new MarketDiscovery(this.deps.predict, this.deps.polymarket);
    const discovered = await discovery.discover();
    await this.record("market_discovery", "discovered BTC markets", {
      predict: discovered.predictMarkets.length,
      polymarket: discovered.polymarketMarkets.length
    });

    const scan = new BtcMarketScanner().scan(discovered);
    await this.record("market_matcher", "strict matcher completed", {
      accepted: scan.accepted.length,
      rejected: scan.rejected.length
    });

    try {
      ensurePolymarketCanOpen(this.deps.polymarketAccount);
    } catch (error) {
      const reason = error instanceof Error ? error.message : "Polymarket account paused";
      await emitMonitorEvent(
        { metrics: this.deps.metrics, alerts: this.deps.alerts },
        botPaused("POLYMARKET_INSUFFICIENT_BALANCE", reason)
      );
      await this.warn(reason);
      return { paused: true, pauseReason: reason, executed: 0 };
    }

    let executed = 0;
    for (const match of scan.accepted) {
      const books = {
        predictYes: await this.deps.predict.getOrderbook(match.predict, "YES"),
        predictNo: await this.deps.predict.getOrderbook(match.predict, "NO"),
        polymarketYes: await this.deps.polymarket.getOrderbook(match.polymarket, "YES"),
        polymarketNo: await this.deps.polymarket.getOrderbook(match.polymarket, "NO")
      };
      const account = this.deps.predictRotator.candidatesFromNext().find((candidate) => isReady(candidate));
      if (!account) {
        await emitMonitorEvent(
          { metrics: this.deps.metrics, alerts: this.deps.alerts },
          {
            eventType: "ALL_PREDICT_ACCOUNTS_UNAVAILABLE",
            severity: "warning",
            reasonCode: "ALL_PREDICT_ACCOUNTS_UNAVAILABLE",
            message: "all Predict accounts are unavailable",
            marketPairId: `${match.predict.venueMarketId}:${match.polymarket.venueMarketId}`
          }
        );
        continue;
      }
      const sized = this.deps.engine.sizeComboA({
        predictYesBook: books.predictYes,
        polymarketNoBook: books.polymarketNo,
        feeRates: this.deps.feeRates,
        limits: {
          selectedPredictFreeBalance: account.availableBalance,
          polymarketAvailableCollateral: this.deps.polymarketAccount.availableCollateral
        }
      });
      if (!sized.executable || !sized.quote) {
        await emitMonitorEvent(
          { metrics: this.deps.metrics },
          tradeRejected("NO_PROFIT_AFTER_BUFFERS", "sized hedge rejected", { reasons: sized.reasons })
        );
        continue;
      }
      const strategyDecision = (this.deps.strategyEngine ?? new StrategyEngine()).evaluate({
        config: this.deps.strategyConfig ?? DEFAULT_STRATEGY_CONFIG,
        pureArbitrage: {
          sizing: sized,
          predictAccountId: account.accountId,
          selectedPredictFreeBalance: account.availableBalance
        }
      });
      if (!strategyDecision.accepted || !strategyDecision.plan) {
        await emitMonitorEvent(
          { metrics: this.deps.metrics },
          tradeRejected("STRATEGY_REJECTED", "strategy layer rejected hedge", { reasons: strategyDecision.reasons })
        );
        continue;
      }
      if (strategyDecision.plan.action !== "OPEN_PURE_ARBITRAGE") {
        await emitMonitorEvent(
          { metrics: this.deps.metrics },
          tradeRejected("STRATEGY_REJECTED", "strategy plan is signal-only and cannot submit orders", {
            action: strategyDecision.plan.action,
            mode: strategyDecision.plan.mode,
            metadata: strategyDecision.plan.metadata
          })
        );
        continue;
      }
      const reserved = this.deps.predictRotator.reserve(account.accountId);
      const predictLeg = strategyDecision.plan.legs.find((leg) => leg.venue === "PREDICT");
      const polymarketLeg = strategyDecision.plan.legs.find((leg) => leg.venue === "POLYMARKET");
      if (!predictLeg || !polymarketLeg) {
        this.deps.predictRotator.release(reserved.accountId);
        await emitMonitorEvent(
          { metrics: this.deps.metrics },
          tradeRejected("STRATEGY_REJECTED", "strategy plan missing required venue legs", { plan: strategyDecision.plan })
        );
        continue;
      }
      const predictOrder: OrderRequest = {
        venue: "PREDICT",
        marketId: match.predict.venueMarketId,
        outcome: predictLeg.outcome,
        side: "BUY",
        orderType: "FOK",
        shares: sized.quote.shares,
        limitPrice: sized.quote.predictLeg.fill.worstPrice ?? sized.quote.predictLeg.fill.averagePrice,
        accountId: reserved.accountId,
        clientOrderId: `predict-${match.predict.venueMarketId}`
      };
      const polymarketOrder: OrderRequest = {
        venue: "POLYMARKET",
        marketId: match.polymarket.venueMarketId,
        outcome: polymarketLeg.outcome,
        side: "BUY",
        orderType: "FOK",
        shares: sized.quote.shares,
        limitPrice: sized.quote.polymarketLeg.fill.worstPrice ?? sized.quote.polymarketLeg.fill.averagePrice,
        accountId: this.deps.polymarketAccount.accountId,
        clientOrderId: `poly-${match.polymarket.venueMarketId}`
      };
      const safetyCheck = validateConservativeOrderPlan({
        predictOrder,
        polymarketOrder,
        predictWorstAcceptablePrice: sized.quote.predictLeg.fill.worstPrice ?? sized.quote.predictLeg.fill.averagePrice,
        polymarketWorstAcceptablePrice: sized.quote.polymarketLeg.fill.worstPrice ?? sized.quote.polymarketLeg.fill.averagePrice
      });
      if (!safetyCheck.ok) {
        this.deps.predictRotator.release(reserved.accountId);
        await emitMonitorEvent(
          { metrics: this.deps.metrics, alerts: this.deps.alerts },
          tradeRejected("UNKNOWN", "execution rejected by conservative order plan", { reasons: safetyCheck.reasons })
        );
        await this.warn(`execution rejected: ${safetyCheck.reasons.join("; ")}`);
        continue;
      }

      let [predictResult, polymarketResult] = this.deps.dryRun
        ? await this.dryRunResults(reserved.accountId, match.predict.venueMarketId, match.polymarket.venueMarketId, sized.quote.shares)
        : await Promise.all([
            this.deps.predict.placeOrder(predictOrder),
            this.deps.polymarket.placeOrder(polymarketOrder)
          ]);
      let safety = this.safetyOutcome(predictResult, polymarketResult, sized.quote.shares);
      const rescueOrder = buildResidualRescueOrder({
        outcome: safety,
        predictOrder,
        polymarketOrder,
        suffix: "rescue"
      });
      if (!this.deps.dryRun && rescueOrder) {
        const rescueResult =
          rescueOrder.venue === "PREDICT"
            ? await this.deps.predict.placeOrder(rescueOrder)
            : await this.deps.polymarket.placeOrder(rescueOrder);
        await this.record("execution_rescue", "residual rescue attempted", {
          venue: rescueOrder.venue,
          status: rescueResult.status,
          filledShares: rescueResult.filledShares.toFixed()
        });
        if (rescueOrder.venue === "PREDICT") {
          predictResult = combineResults(predictResult, rescueResult);
        } else {
          polymarketResult = combineResults(polymarketResult, rescueResult);
        }
        safety = this.safetyOutcome(
          predictResult,
          polymarketResult,
          sized.quote.shares,
          rescueResult.filledShares.gte(safety.residualShares) ? "succeeded" : "failed"
        );
      }
      await this.record("execution", "execution results", {
        predict: predictResult.status,
        polymarket: polymarketResult.status,
        safetyState: safety.state,
        safetyActions: safety.actions
      });
      await emitMonitorEvent(
        { metrics: this.deps.metrics, alerts: this.deps.alerts },
        orderSubmitted({
          hedgeId: `${match.predict.venueMarketId}:${match.polymarket.venueMarketId}`,
          orderId: predictResult.exchangeOrderId ?? predictResult.clientOrderId,
          venue: "PREDICT"
        })
      );
      await emitMonitorEvent(
        { metrics: this.deps.metrics, alerts: this.deps.alerts },
        orderSubmitted({
          hedgeId: `${match.predict.venueMarketId}:${match.polymarket.venueMarketId}`,
          orderId: polymarketResult.exchangeOrderId ?? polymarketResult.clientOrderId,
          venue: "POLYMARKET"
        })
      );
      this.finishPredictAccount(reserved.accountId, match.predict, predictOrder.outcome, predictResult, safety);
      executed += 1;
      if (safety.state === "PAUSED" || safety.actions.includes("PAUSE_NEW_OPENINGS")) {
        const reason = safety.pauseReason ?? (safety.reasons.join("; ") || "execution safety pause");
        await emitMonitorEvent(
          { metrics: this.deps.metrics, alerts: this.deps.alerts },
          botPaused(safety.pauseReason === "residual rescue failed" ? "RESCUE_FAILED" : "UNHEDGED_RESIDUAL", reason, {
            safetyState: safety.state,
            actions: safety.actions,
            reasons: safety.reasons
          })
        );
        await this.warn(reason);
        return { paused: true, pauseReason: reason, executed };
      }
    }

    return { paused: false, executed };
  }

  private async dryRunResults(
    accountId: string,
    predictMarketId: string,
    polymarketMarketId: string,
    shares: D
  ): Promise<[OrderResult, OrderResult]> {
    const predict: OrderResult = {
      venue: "PREDICT",
      clientOrderId: `dry-predict-${predictMarketId}`,
      status: "dry_run",
      filledShares: shares,
      averagePrice: ZERO
    };
    const polymarket: OrderResult = {
      venue: "POLYMARKET",
      clientOrderId: `dry-poly-${polymarketMarketId}`,
      status: "dry_run",
      filledShares: shares,
      averagePrice: ZERO,
      raw: { accountId }
    };
    return [predict, polymarket];
  }

  private async record(eventType: string, message: string, data?: Record<string, unknown>) {
    await this.deps.audit?.record(auditEvent(eventType, message, data));
  }

  private async warn(message: string) {
    await this.record("alert", message);
    await this.deps.alerts?.send(alert("warning", message));
  }

  private safetyOutcome(
    predictResult: OrderResult,
    polymarketResult: OrderResult,
    requestedShares: D,
    rescueAttempt: "none" | "succeeded" | "failed" = "none"
  ): ExecutionSafetyOutcome {
    return reconcileExecutionSafety({
      predict: confirmOrderFill({ placement: predictResult, requestedShares, requireRestAndWs: false }),
      polymarket: confirmOrderFill({ placement: polymarketResult, requestedShares, requireRestAndWs: false }),
      requestedShares,
      policy: this.deps.safetyPolicy ?? {
        maxUnhedgedMs: 3000,
        cancelUnfilledOrders: true,
        postTradeReconcileRequired: false,
        pauseOnUnhedgedResidual: true
      },
      nowMs: Date.now(),
      rescueAttempt
    });
  }

  private finishPredictAccount(
    accountId: string,
    market: BinaryMarketSpec,
    outcome: HeldPosition["outcome"],
    predictResult: OrderResult,
    safety: ExecutionSafetyOutcome
  ): void {
    if (this.deps.dryRun) {
      this.deps.predictRotator.release(accountId);
      return;
    }
    if (!this.deps.dryRun && predictResult.filledShares.gt(0)) {
      this.deps.predictRotator.markHeld(accountId, {
        marketId: market.venueMarketId,
        outcome,
        shares: predictResult.filledShares,
        costBasis: predictResult.filledShares.mul(predictResult.averagePrice),
        oracleStatus: "PENDING_UMA_FINALITY",
        redeemed: false,
        eventEndTs: parseDate(market.windowEndUtc),
        heldSince: new Date(),
        heldReason: "opened by execution coordinator"
      });
      return;
    }
    if (safety.actions.includes("RELEASE_PREDICT_ACCOUNT_LOCK") || predictResult.filledShares.eq(0)) {
      this.deps.predictRotator.release(accountId);
    }
  }
}

function parseDate(value: string | undefined): Date | undefined {
  if (!value) return undefined;
  const parsed = new Date(value);
  return Number.isNaN(parsed.getTime()) ? undefined : parsed;
}

function combineResults(left: OrderResult, right: OrderResult): OrderResult {
  const filledShares = left.filledShares.plus(right.filledShares);
  const gross = left.filledShares.mul(left.averagePrice).plus(right.filledShares.mul(right.averagePrice));
  return {
    ...left,
    status: filledShares.gt(0) ? "matched" : right.status,
    filledShares,
    averagePrice: filledShares.gt(0) ? gross.div(filledShares) : ZERO,
    raw: {
      first: left.raw,
      rescue: right.raw
    }
  };
}
